oci/archive/oci_src.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/pkg/errors"
+)
+
+type ociArchiveImageSource struct {
+	ref         ociArchiveReference
+	unpackedSrc types.ImageSource
+	tempDir     string
+}
+
+var tempDirRef *tempDirOciRef
+
+// newImageSource returns an ImageSource for reading from an existing directory.
+// newImageSource untars the file and saves it in a temp directory
+func newImageSource(ref ociArchiveReference, ctx *types.SystemContext, requestedManifestMIMETypes []string) (types.ImageSource, error) {
+	if tempDirRef == nil {
+		if err := createUntarTempDir(ref); err != nil {
+			return nil, errors.Wrap(err, "error creating temp directory")
+		}
+	}
+	unpackedSrc, err := tempDirRef.ociRef.NewImageSource(ctx, requestedManifestMIMETypes)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return nil, err
+	}
+	return &ociArchiveImageSource{ref: ref,
+		unpackedSrc: unpackedSrc,
+		tempDir:     tempDirRef.tempDirectory}, nil
+}
+
+// LoadManifestDescriptor loads the manifest
+func LoadManifestDescriptor(imgRef types.ImageReference) (imgspecv1.Descriptor, error) {
+	if tempDirRef == nil {
+		if err := createUntarTempDir(imgRef.(ociArchiveReference)); err != nil {
+			return imgspecv1.Descriptor{}, errors.Wrap(err, "error creating temp directory")
+		}
+	}
+	descriptor, err := ocilayout.LoadManifestDescriptor(tempDirRef.ociRef)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return imgspecv1.Descriptor{}, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return imgspecv1.Descriptor{}, errors.Wrap(err, "error loading manifest")
+	}
+	return descriptor, nil
+}
+
+// Reference returns the reference used to set up this source.
+func (s *ociArchiveImageSource) Reference() types.ImageReference {
+	return s.ref
+}
+
+// Close removes resources associated with an initialized ImageSource, if any.
+// Close deletes the temporary directory at dst
+func (s *ociArchiveImageSource) Close() error {
+	if err := s.unpackedSrc.Close(); err != nil {
+		if err := deleteTempDir(); err != nil {
+			return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return errors.Wrapf(err, "error closing image source")
+	}
+	if err := deleteTempDir(); err != nil {
+		return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+	}
+	return nil
+}
+
+// GetManifest returns the image's manifest along with its MIME type
+// (which may be empty when it can't be determined but the manifest is available).
+func (s *ociArchiveImageSource) GetManifest() ([]byte, string, error) {
+	return s.unpackedSrc.GetManifest()
+}
+
+func (s *ociArchiveImageSource) GetTargetManifest(digest digest.Digest) ([]byte, string, error) {
+	valByte, valString, err := s.unpackedSrc.GetTargetManifest(digest)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
```

Comment:
Fixed. I have cleanups in `newImageDestination` and `newImageSource` because it is possible for a failure to occur there after the temp directory is created.